Test with latest setuptools 75.8.2, and with pip 25.0.1.

mauritsvanrees opened the pull request:
I suspect 75.8.1 from yesterday will break for us due to changes in wheel names. Indeed this seems the case locally.  With 75.8.0 from early January I expect all is still fine.

[Update: original title was "Test with latest setuptools 75.8.0 and 75.8.1, and with pip 25.0.1."  Now I have updated the PR to test with latest 75.8.2.]

mauritsvanrees commented:
Actually, our cron tests with 'setuptools HEAD' started [failing](https://github.com/buildout/buildout/actions/runs/13104747791/job/36557695986#step:5:235) three weeks ago, with tracebacks like this:

```
Got zc-recipe-egg 2.0.7.
While:
  Installing.
  Getting section test.
  Initializing section test.
  Loading zc.buildout recipe entry zc.recipe.egg:default.

An internal error occurred due to a bug in either zc.buildout or in a
recipe being used:
Traceback (most recent call last):
  File "/home/runner/work/buildout/buildout/src/zc/buildout/buildout.py", line 2283, in main
    getattr(buildout, command)(args)
  File "/home/runner/work/buildout/buildout/src/zc/buildout/buildout.py", line 726, in install
    [self[part]['recipe'] for part in install_parts]
  File "/home/runner/work/buildout/buildout/src/zc/buildout/buildout.py", line 726, in <listcomp>
    [self[part]['recipe'] for part in install_parts]
     ~~~~^^^^^^
  File "/home/runner/work/buildout/buildout/src/zc/buildout/buildout.py", line 1359, in __getitem__
    options._initialize()
  File "/home/runner/work/buildout/buildout/src/zc/buildout/buildout.py", line 1471, in _initialize
    self.initialize()
  File "/home/runner/work/buildout/buildout/src/zc/buildout/buildout.py", line 1477, in initialize
    recipe_class = _install_and_load(reqs, 'zc.buildout', entry, buildout)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/runner/work/buildout/buildout/src/zc/buildout/buildout.py", line 1431, in _install_and_load
    return pkg_resources.load_entry_point(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/runner/work/buildout/buildout/sandbox/src/setuptools/pkg_resources/__init__.py", line 535, in load_entry_point
    return get_distribution(dist).load_entry_point(group, name)
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/home/runner/work/buildout/buildout/sandbox/src/setuptools/pkg_resources/__init__.py", line 527, in get_distribution
    dist = get_provider(dist)
           ^^^^^^^^^^^^^^^^^^
  File "/home/runner/work/buildout/buildout/sandbox/src/setuptools/pkg_resources/__init__.py", line 414, in get_provider
    return working_set.find(moduleOrReq) or require(str(moduleOrReq))[0]
                                            ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/runner/work/buildout/buildout/sandbox/src/setuptools/pkg_resources/__init__.py", line 1061, in require
    needed = self.resolve(parse_requirements(requirements))
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/runner/work/buildout/buildout/sandbox/src/setuptools/pkg_resources/__init__.py", line 888, in resolve
    dist = self._resolve_dist(
           ^^^^^^^^^^^^^^^^^^^
  File "/home/runner/work/buildout/buildout/sandbox/src/setuptools/pkg_resources/__init__.py", line 929, in _resolve_dist
    raise DistributionNotFound(req, requirers)
pkg_resources.DistributionNotFound: The 'zc.recipe.egg' distribution was not found and is required by the application
```

And the main [setuptools change](https://github.com/pypa/setuptools/pull/4766) was made three weeks ago.  So we could have known this... :-/

Similar problem reported in [`plone.autoinclude`](https://github.com/plone/plone.autoinclude/issues/25) today.  (So maybe `z3c.autoinclude` is also affected.)

mauritsvanrees commented:
I have updated the PR to test with latest 75.8.2, which avoids some problems.  [This job](https://github.com/buildout/buildout/actions/runs/13564894216/job/37915701161?pr=677) with 75.8.2 at least succeeds.  Further tests are in progress.

I have added a feature to the `ls` testing method, which now accepts a keyword argument `lowercase_and_sort_output`.  Default is False, and I set it to true in one test.  See commit daa5bc79ae9728319f53ae6aa0aec27a7ee7bb56.

mauritsvanrees commented:
The script tests for this PR were failing with `TypeError: ('Expected str, Requirement, or Distribution', None)`.  In this test we are installing `zest.releaser` and then we get its distribution and ask for the entry points, so we can create scripts.  But the `zest.releaser` distribution is not found.  Instead we can find it by its canonicalised/normalised name: `zest-releaser`.  (Or also via `zest_releaser`, but the dashes seem better.).  I fix that in the latest commit b7550406600612c415d9074dedb19807c11d0cd8.

FYI, locally this test actually passed when I updated the sandbox environment to use latest `setuptools` 75.8.2.  For me the venv had 65.5.0 by default, and I tried a few, I think up to and including 75.8.1, and they all failed.  With this latest commit, it works on all versions.

icemac commented:
I've also seen both of these errors (`zest.releaser` and `zc.recipe.egg`) on our company Jenkins. I solved it by updating to `setuptools == 75.8.2` – The failures seemed to have happened over night, I did not find any connection to updated version numbers.

icemac commented on src/zc/buildout/tests/easy_install.txt:
```diff
@@ -258,14 +258,16 @@ Let's enable server logging to check that the lower case file is downloaded.
     GET 200 /demoneeded-1.1.zip
 
 Let's check that the uppercase dist is installed.
+setuptools 75.8 1+ reports the name in all lowercase, earlier versions showed it in uppercase.
```
```suggestion
setuptools 75.8.1+ reports the name in all lowercase, earlier versions showed it in uppercase.
```

icemac reviewed (approved):
LGTM. (Would be nice to get it merged and released quickly.)

There are two required jobs which did not run – any idea what to do there?

mauritsvanrees commented:
> LGTM. (Would be nice to get it merged and released quickly.)
> 
> There are two required jobs which did not run – any idea what to do there?

Ah, I see.  We require windows/mac-latest on Py 3.10 with setuptools 75.6.0.  These jobs no longer exist, I have replaced them with jobs testing 75.8.2.  Those two passed.
I have updated the branch protection rule to require the two new jobs.  Now it should be fine.

mauritsvanrees commented:
Thank you for reviewing!

The build is green and I have merged.

I have released [4.1 on PyPI](https://pypi.org/project/zc.buildout/4.1/).
